### ISSUE TITLE
Product API updates error

### ISSUE THREAD

**Issue by jakwinkler** [magento/magento2#12508]
This bug is related to updating product data by SOAP / REST API.

### Preconditions
1. Magento 2.1.7
2. Simple / Configurable product save / update by API

### Steps to reproduce
1. Create a simple / configurable product manually. Set it status disabled.
2. Create a SOAP request to update that product to: **change product status** to enabled and **assign product to some categories**, for example:
```
<item>
	<attributeCode>category_ids</attributeCode>
	<value>
	       <xsd:array>46</xsd:array>
	       <xsd:array>48</xsd:array>
	       <xsd:array>59</xsd:array>
       </value>
</item>
```
3. It is important to have catgories in SOAP / REST request.

### Expected result

1. Save product, change status and assign it to new / different categories.

### Actual result

1. `Unable to save productException: Notice: Array to string conversion`
in [...]vendor/magento/module-catalog/Model/Product.php on line 2290

In product Model $this->getCategoryIds() returns associative array with array key equal to the one defined in schema **xsd:array** 
Basically when using SOAP, getCategoryIds returns 2 dimensional array, when it should only return one dimensional array with category IDs

**Comment by jakwinkler** [comment]
getCategoryIds() returns following data structure:
```
array(1) {
  ["array"]=>
  array(6) {
    [0]=>
    string(2) "46"
    [1]=>
    string(2) "48"
    [2]=>
    string(2) "59"
    [3]=>
    string(2) "82"
    [4]=>
    string(2) "85"
    [5]=>
    string(2) "87"
  }
}
``` 
for data above

**Comment by jakwinkler** [comment]
Here is how I've solved this temporary for the Client, but in my opinion SOAP XML mapping should be a place where final fix takes place and always return 1 dimensional array for getCategoryIds():
```
if (is_array($categoryId)) {
                    foreach ($categoryId as $apiCategoryId) {
                        $identities[] = self::CACHE_PRODUCT_CATEGORY_TAG . '_' . $apiCategoryId;
                    }
                } else {
                    $identities[] = self::CACHE_PRODUCT_CATEGORY_TAG . '_' . $categoryId;
                }
```
vendor/magento/module-catalog/Model/Product.php, **getIdentities()** method

**Comment by magento-engcom-team** [comment]
Hello @qsolutions-pl. Thanks for reporting. To make changes in product You aimed to achieve we have tried the following requests.

SOAP
```
http://{yourmagentosite}/soap/default?services=catalogProductRepositoryV1
<?xml version="1.0" encoding="UTF-8"?>
<soap:Envelope xmlns:def="http://{yourmagentosite}/soap/default?services=catalogProductRepositoryV1" xmlns:soap="http://www.w3.org/2003/05/soap-envelope">
	<soap:Body>
		<def:catalogProductRepositoryV1SaveRequest>
            <product>
				<sku>{Your product's SKU}</sku>
				<status>1</status>
              	<customAttributes>
                    <item>
                        <attributeCode>category_ids</attributeCode>
                        <value>
                            <array>46</array>
			    <array>48</array>
                            <array>59</array>
                        </value>  
                    </item>
                </customAttributes>  
            </product>
		</def:catalogProductRepositoryV1SaveRequest>
	</soap:Body>
</soap:Envelope>
```

REST
```
Method PUT http://{yourmagentosite}/rest/V1/products/{Your product's SKU}
body:
{
    "product":{
        "status": 1,
        "custom_attributes": [
            {
                    "attribute_code": "category_ids",
                    "value": ["46", "48", "59"]
            }
        ]
    }
}
```

Seems they both work correctly and no changes to code are needed. Please tell us whether they would work for You. Thank You.

**Comment by jakwinkler** [comment]
The problem occurs only when changing product status.

**Comment by magento-engcom-team** [comment]
@qsolutions-pl please specify which status do You mean Stock Status (In Stock - Out of Stock) or Enabled Product (Yes-No)?

**Comment by magento-engcom-team** [comment]
@qsolutions-pl, we are closing this issue due to inactivity. If you'd like to update it, please reopen the issue.
